weblate/vcs/tests/test_vcs.py

```diff
@@ -435,19 +434,406 @@ def add_remote_commit(self, conflict=False, rename=False):
+        )
+        mock_pr_response = self._mock_response(
+            status=200, json_data={"message": "Some error"}
+        )
+        mock_post.side_effect = [mock_fork_response, mock_pr_response]
+
+        mock_put_patcher = patch("requests.put")
+        mock_put = mock_put_patcher.start()
+        mock_put.return_value = self._mock_response(
+            status=200, json_data={"web_url": "https://gitlab.com/test/test"},
+        )
+
+        mock_get_patcher = patch("requests.get")
+        mock_get = mock_get_patcher.start()
+        mock_get.side_effect = self.get_empty_side_effect
+
+        with self.assertRaises(RepositoryException):
+            super().test_push(branch)
+        mock_get_patcher.stop()
+        mock_put_patcher.stop()
+        mock_post_patcher.stop()
+        mock_push_to_fork.stop()
+
+    def test_pull_request_exists(self, branch=""):
+        self.repo.component.repo = "https://gitlab.com/WeblateOrg/test.git"
+
+        # Patch push_to_fork() function because we don't want to actually
+        # make a git push request
+        mock_push_to_fork_patcher = patch(
+            "weblate.vcs.git.GitMergeRequestBase.push_to_fork"
+        )
+        mock_push_to_fork = mock_push_to_fork_patcher.start()
+        mock_push_to_fork.return_value = ""
+
+        # Check that it doesn't raise error when pull request already exists
+        mock_post_patcher = patch("requests.post")
+        mock_post = mock_post_patcher.start()
+        mock_fork_response = self._mock_response(
+            status=200,
+            json_data={
+                "ssh_url_to_repo": "git@gitlab.com:test/test.git",
+                "_links": {"self": "https://gitlab.com/api/v4/projects/20227391"},
+            },
+        )
+        mock_pr_response = self._mock_response(
+            status=200,
+            json_data={"message": ["Another open merge request already exists"]},
+        )
+        mock_post.side_effect = [mock_fork_response, mock_pr_response]
+
+        mock_put_patcher = patch("requests.put")
+        mock_put = mock_put_patcher.start()
+        mock_put.return_value = self._mock_response(
+            status=200, json_data={"web_url": "https://gitlab.com/test/test"},
+        )
+
+        mock_get_patcher = patch("requests.get")
+        mock_get = mock_get_patcher.start()
+        mock_get.side_effect = self.get_empty_side_effect
+
+        super().test_push(branch)
+        mock_get_patcher.stop()
+        mock_put_patcher.stop()
+        mock_post_patcher.stop()
+        mock_push_to_fork.stop()
+
+
+class VCSGerritTest(VCSGitTest):
     _class = GitWithGerritRepository
     _vcs = "git"
     _sets_push = True
 
+    def add_remote_commit(self, conflict=False, rename=False):
+        # Use Git to create changed upstream repo
+        backup = self._class
+        self._class = GitRepository
+        try:
+            super().add_remote_commit(conflict, rename)
+        finally:
+            self._class = backup
```

Comment:
I'd prefer to reuse this code instead of having it implemented in multiple test classes.

Reply:
I will create an intermediate class then from which I will inherit VCSGithubTest, VCSGitlabTest and VCSGerritTest.